stablehlo/reference/tests/StablehloInterpreterRunner.cpp

```diff
@@ -0,0 +1,122 @@
+#include "stablehlo/reference/Tensor.h"
+
+using namespace llvm;
+using namespace mlir;
+
+namespace {
+
+/// Parses the source program represented by the memory buffer, runs the
+/// interpreter on the parsed function, and finally prints the evaluated value
+/// to os.
+static LogicalResult processBuffer(std::unique_ptr<MemoryBuffer> chunkBuffer,
+                                   raw_ostream &os) {
+  // Registers Dialects and sets up the MLIR context.
+  DialectRegistry registry;
+  registry.insert<func::FuncDialect>();
+  registry.insert<stablehlo::StablehloDialect>();
+  MLIRContext context(registry);
+  context.loadAllAvailableDialects();
+
+  SourceMgr sourceMgr;
+  sourceMgr.AddNewSourceBuffer(std::move(chunkBuffer), SMLoc());
+  SourceMgrDiagnosticHandler sourceMgrHandler(sourceMgr, &context);
+
+  // Prepare the input buffer.
+  ParserConfig config(&context);
+  OwningOpRef<ModuleOp> module(parseSourceFile<ModuleOp>(sourceMgr, config));
+  if (!module) return failure();
+
+  for (auto funcOp : module->getBodyRegion().getOps<mlir::func::FuncOp>()) {
+    os << "\nEvaluated results of function: " << funcOp.getSymName() << "\n";
+
+    // Run the test model.
+    auto results = mlir::stablehlo::eval(funcOp, {});
+    if (!(bool)results) toString(results.takeError());
+
+    // Dump the results.
+    for(auto &result : *results) result.print(os);
+  }
+
+  return success();
+}
+
+LogicalResult InterpreterMain(int argc, char **argv) {
+  static cl::opt<std::string> inputFilename(
+      cl::Positional, cl::desc("<input file>"), cl::init("-"));
+
+  static cl::opt<std::string> outputFilename("o", cl::desc("Output filename"),
+                                             cl::value_desc("filename"),
+                                             cl::init("-"));
+
+  static cl::opt<bool> splitInputFile(
+      "split-input-file",
+      cl::desc("Split the input file into pieces and process each "
+               "chunk independently"),
+      cl::init(false));
+
+  InitLLVM y(argc, argv);
+  cl::ParseCommandLineOptions(argc, argv, "StableHLO interpreter test runner");
+
+  // Set up the input/output files.
+  std::string errorMessage;
+  auto file = openInputFile(inputFilename, &errorMessage);
+  if (!file) {
+    llvm::errs() << errorMessage << "\n";
+    return failure();
+  }
+
+  auto output = openOutputFile(outputFilename, &errorMessage);
+  if (!output) {
+    llvm::errs() << errorMessage << "\n";
+    return failure();
+  }
+
+  // Process the input buffer.
+  if (failed(splitAndProcessBuffer(std::move(file), processBuffer, output->os(),
+                                   splitInputFile,
+                                   /*insertMarkerInOutput=*/true)))
+    return failure();
+
+  // Keep the output file if the invocation of InterpreterMain was successful.
+  output->keep();
+  return success();
+}
+
+}  // namespace
+
+int main(int argc, char **argv) { return failed(InterpreterMain(argc, argv)); }
```

Comment:
Can we use https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Tools/mlir-translate/MlirTranslateMain.h for this? It handles a bunch of boilerplate for us.